### PR TITLE
nmcli: 'slave_type' and 'may_fail6' implemented, idempotency fixes

### LINKED ISSUE #473: Add ability in nmcli module to bind vlan interface to a bridge interface.
##### SUMMARY

In the `nmcli` module, add a `master` and `slave-type` parameter for vlan interface types to allow them to be enslaved onto bridge interfaces.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

nmcli

##### ADDITIONAL INFORMATION

I'm currently trying to setup a bunch of different network interfaces for an Openstack deployment in my home lab. Per the [Openstack documentation](https://docs.openstack.org/openstack-ansible/latest/user/network-arch/example.html#single-interface-or-bond), I'm need to create three bridge interfaces that are all on separate vlans. I'm trying to use the `nmcli` module to do this, however it seems currently impossible given the list of possible options in the [documentation](https://docs.ansible.com/ansible/latest/modules/nmcli_module.html). 

From what I can see based on [this snippet](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/sec-vlan_on_bond_and_bridge_using_the_networkmanager_command_line_tool_nmcli) of RedHat documentation, what I'd like is for the following playbook:

```yaml
-name: Add br-mgmt interface
  nmcli:
    state: present
    conn_name: br-mgmt
    ifname: br-mgmt
    type: bridge
- name: Create vlan 100 interface and bind to br-mgmt
  nmcli:
    state: present
    conn_name: em1.100
    ifname: em1.100
    type: vlan
    vlanid: 100
    vlandev: em1
    master: br-mgmt
    slave-type: bridge
```

To execute something like the following:

```
nmcli connection add type bridge con-name br-mgmt ifname br-mgmt
nmcli connection add type vlan con-name em1.100 ifname em1.100 dev em1 id 100 master br-mgmt slave-type bridge
```

It's the functionality of appending the `master br-mgmt slave-type bridge` to the command that would need to be implemented in the `nmcli` module.

### LINKED ISSUE #4387: nmcli does not allow editing mtu on VLAN interfaces
### Summary

When trying to edit the mtu parameter on a VLAN interface, nmcli always reports that nothing needs to be done, even though the specified mtu value is different than the current mtu for the VLAN Interface.

This appears to be caused by the function:

```python
    def mtu_conn_type(self):
        return self.type in (
            'dummy',
            'ethernet',
            'team-slave',
        )
```

in `nmcli.py` which fails to recognize the `vlan` interface as a valid layer 2 interface type on which you can set the MTU.

### Issue Type

Bug Report

### Component Name

nmcli

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.12.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.10/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.10.3 (main, Mar 17 2022, 00:10:53) [GCC 8.3.0]
  jinja version = 3.0.3
  libyaml = True
```

### Community.general Version

```console
$ ansible-galaxy collection list community.general

# /root/.ansible/collections/ansible_collections
Collection        Version
----------------- -------
community.general 4.6.0

# /usr/local/lib/python3.10/site-packages/ansible_collections
Collection        Version
----------------- -------
community.general 4.6.0
```

### Configuration

```console
$ ansible-config dump --only-changed
HOST_KEY_CHECKING(/etc/ansible/ansible.cfg) = False
```

### OS / Environment

Rocky Linux 8.5

### Steps to Reproduce

```yaml
 - name: Set VLAN 96 network interface MTU to 1500 after creation
    community.general.nmcli:
      conn_name:  "enp1s0f0.96"
      type:              "vlan"
      vlandev:        "enp1s0f0"
      vlanid:           96
      ip4:               "1.2.3.4"
      mtu:             9000
      state:           "present"
```

### Expected Results

```console
$ nmcli conn show enp1s0f096
```

should report:

```console
802-3-ethernet.mtu:                     9000
```

but instead Ansible reports that nothing needs to be changed and the MTU remains at the default for the base interface, 1500.

### Actual Results

```console (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### LINKED ISSUE #4683: nmcli bridge not idempotent
### Summary

When configuring a bridge with stp disabled, the bridge's stp priority changes from 32768 to 128 on every run (also the type conversion is broken, but that's more a cosmetic issue).

### Issue Type

Bug Report

### Component Name

nmcli

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.12.5]
  config file = /home/phil/projects/ansible/ansible.cfg
  configured module search path = ['/home/phil/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.10/site-packages/ansible
  ansible collection location = /home/phil/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.10.4 (main, Mar 25 2022, 00:00:00) [GCC 11.2.1 20220127 (Red Hat 11.2.1-9)]
  jinja version = 3.0.1
  libyaml = True
```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general
# /home/phil/.ansible/collections/ansible_collections
Collection        Version
----------------- -------
community.general 4.8.1
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed
(empty)
```

### OS / Environment

OracleLinux 8.6

### Steps to Reproduce

```yaml (paste below)
    - community.general.nmcli:
        conn_name: br0 
        ifname: br0 
        type: bridge
        stp: no                                                                                                        
        ip4: '10.0.200.1/24'
        state: present
    - community.general.nmcli:
        conn_name: eth1
        ifname: eth1
        type: bridge-slave
        master: br0 
        hairpin: false
        stp: no
        state: present
```

### Expected Results

no changes on consecutive runs

### Actual Results

```console (paste below)
TASK [community.general.nmcli] ****************************************************************************************
--- before
+++ after
@@ -1,9 +1,9 @@
 {
-    "bridge.ageing-time": "300",
-    "bridge.forward-delay": "15",
-    "bridge.hello-time": "2",
-    "bridge.max-age": "20",
-    "bridge.priority": "32768",
+    "bridge.ageing-time": 300,
+    "bridge.forward-delay": 15,
+    "bridge.hello-time": 2,
+    "bridge.max-age": 20,
+    "bridge.priority": 128,
     "bridge.stp": "no",
     "connection.autoconnect": "yes",
     "connection.interface-name": "br0",
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by heaveaxy**
New options: 
- `slave_type` - allowing to implement slave connections of non-ethernet types, eg. bond to bridge or vlan to bridge.
- `may_fail6` - allowing to control may-fail option of ipv6 that already implemented for ipv4

Bugfixes: 
- task always changed when changing bridge's stp.priority; nmcli accepts option when stp disabled but keeps default value
- idempotency issue when changing connection's ipv4.may-fail; nmcli accepts option when method is 'disabled' but keeps default value
  some refactoring of code in part of method names

##### SUMMARY
* Adding `slave_type` option have discussed in several issues:
  * #2943 
  * Fixes #473

  Without it module makes impossible to create non-ethernet interfaces (eg bond, vlan) attached to bridges. From nmcli side this options already deprecated.
* Adding `may_fail6` for IPv6. This option is exists for ipv4, but not for ipv6.
  There is an idempotency issue with `may_fail4`. If `method4` was `disabled`, this option has no effect - nmcli accepts it, but keeps default value. Task marked changed always because of `diff` with default value and passed to module. Now option `ipv4.may-fail` (`ipv6.may-fail` too) participate in configuration process only when user defined `method4` (`method6`) as non-`disabled` and non-`ignore` (for IPv6).
* Same thing was found in `priority` (for bridge types). When `bridge.stp` is set to `no`, `priority` has no effect. Nmcli keeps default value. Related issues:
  * Fixes #4683
  For now, `bridge.priority` participate in configure only when user defined `stp` as `true`.

* Changing MTU for `vlan` interface types was impossible because this logic just was not provided. I define `vlan` type in list of mtu_interfaces.
  * Fixes #4387

* Some small refactoring is made to make code is understandable (rename bool methods with prefix `is_`).

##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Refactoring Pull Request
- Test Pull Request

##### COMPONENT NAME
nmcli

##### ADDITIONAL INFORMATION
Due to options changed and adding new values it reqired to update tests. I did.

Also I have add warning when task run with `type`=`bridge-slave`. In my opinion this `type` is very problematic due to it has non-obviousness results.

### CODE REVIEW

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - fixed idempotency issue when modifying bridge connections - module forces default value of bridge.priority to nmcli; when bridge.stp is disabled nmcli ignores is and keep default (https://github.com/ansible-collections/community.general/issues/3216, https://github.com/ansible-collections/community.general/issues/4683)
```
Comment (felixfontein) [inline changelogs/fragments/3216-nmcli-idempotency-fixes.yml:2]: Please note that fragments should look like this:
```suggestion
  - nmcli - fixed idempotency issue when modifying bridge connections. Module forces default value of bridge.priority to nmcli; when bridge.stp is disabled nmcli ignores is and keep default (https://github.com/ansible-collections/community.general/issues/3216, https://github.com/ansible-collections/community.general/issues/4683).
```

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+  - fixed idempotency issue with 'may_fail4=false' passes when method4 is 'disabled'; in this case nmcli ignores change and keeps 'yes'.
```
Comment (felixfontein) [inline changelogs/fragments/3216-nmcli-idempotency-fixes.yml:3]: Every fragment must include a link to PR and/or issue. Also never use `yes` or `no` for booleans, always use `true` and `false`.

File: changelogs/fragments/473-nmcli-slave-type-implemented.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - nmcli - new module option 'slave_type' added to allow creation of various types of slave devices (https://github.com/ansible-collections/community.general/issues/473)
```
Comment (felixfontein) [inline changelogs/fragments/473-nmcli-slave-type-implemented.yml:2]: Please move new features to another PR since bugfixes are backported to other stable branches than new features. (Also having different features in different PRs makes this easier to review.)

File: plugins/modules/nmcli.py [hunk 4]
```diff
@@ -1738,7 +1803,7 @@ def connection_options(self, detect_change=False):
         return options
 
     @property
-    def ip_conn_type(self):
+    def is_ip_conn_type(self):
```
Comment (felixfontein) [inline plugins/modules/nmcli.py:1806]: Please note that changing code style for code you are not maintaining is never a good idea without asking first.

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution. I've added some comments below.

Also please remove all `__init__.py` files from this PR.

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml (hunk 1 above)
Comment (heaveaxy) [reply changelogs/fragments/3216-nmcli-idempotency-fixes.yml:2]: Ok, I will correct all fragments

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml (hunk 2 above)
Comment (heaveaxy) [reply changelogs/fragments/3216-nmcli-idempotency-fixes.yml:3]: In this case 'yes' - is a value provided by nmcli. I think I will paraphrase to make it clearer.
About PR, if I understand correct, to link fragment to PR it need create PR firstable then commit fragment change with actual link?

File: plugins/modules/nmcli.py (hunk 4 above)
Comment (heaveaxy) [reply plugins/modules/nmcli.py:1806]: Agrees this. I should rollback it and do it correct way. Can you describe please, how to determine who maintaining code of one or another module?

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml (hunk 2 above)
Comment (felixfontein) [reply changelogs/fragments/3216-nmcli-idempotency-fixes.yml:3]: > About PR, if I understand correct, to link fragment to PR it need create PR firstable then commit fragment change with actual link?

Yes, or you do some guessing (PR numbers are rather predictable). Doing it in two commits is easier though.

File: plugins/modules/nmcli.py (hunk 4 above)
Comment (felixfontein) [reply plugins/modules/nmcli.py:1806]: Usually you can see whom the bot automatically pings when creating a new issue or PR. In the case of nmcli, there is no active maintainer, but there are a lot of individual contributions from several persons.

In general the best approach is to first create an issue explaining what you think would be better, see whether the bot tags someone, then wait to see whether these persons are still active / interested, and if there is no reaction try tagging persons who made changes to that piece of code in the past (and thus seem to be interested in it as well). This is a lengthy process, but it will avoid unnecessary work.

Comment (heaveaxy) [reply plugins/modules/nmcli.py:1806]: Good. Thanks you for explanation!

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml (hunk 1 above)
Comment (heaveaxy) [reply changelogs/fragments/3216-nmcli-idempotency-fixes.yml:2]: I corrected all fragments in new PRs

File: changelogs/fragments/3216-nmcli-idempotency-fixes.yml (hunk 2 above)
Comment (heaveaxy) [reply changelogs/fragments/3216-nmcli-idempotency-fixes.yml:3]: I add links to fragments in PR that opened separately of this.

File: changelogs/fragments/473-nmcli-slave-type-implemented.yml (hunk 3 above)
Comment (heaveaxy) [reply changelogs/fragments/473-nmcli-slave-type-implemented.yml:2]: I split all these changes to different PRs

**Comment by heaveaxy** [comment]
Closing this PR because all this changes was sent via another PRs:
#6104
#6105
#6106
#6107
#6108
